### PR TITLE
Connect to backend with workaround

### PR DESCRIPTION
**Pull request by Arif-Khalid**
### Changes made
* Connected frontend login to backend auth0 login
  * Due to fault of Auth0 being unable to allow origin when redirecting, workaround of setting token in url
  * Edited routes to accept token param in url
  * Used session storage to store authentication token
* Connected frontend product querying to backend
  * Use proxy to enable cookies to be sent from server to client
  * Add token from url in frontend to cookie so require_login decorator in backend can be applied
  * Added missing properties from frontend Products appropriately

### CODE REVIEW

File: frontend/src/context/auth-context.tsx [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
   logOut: () => void
+  authenticate: (token: string) => void
 }
 
 const AuthContext = createContext(null as AuthContextType | null)
 
 export const AuthProvider = ({ children }: { children: JSX.Element }) => {
-  const [token, setToken] = useState(null as string | null)
+  const [token, setToken] = useState(
+    sessionStorage.getItem(AUTH_TOKEN_STORAGE_KEY) as string | null
```
Comment (nknguyenhc) [inline frontend/src/context/auth-context.tsx]: `sessionStorage.getItem` return type is `string | null`, so there is no need for typecasting here.

File: frontend/src/hooks/useProducts.ts [hunk 2]
```diff
@@ -7,96 +7,87 @@ import { useLoading } from '../context/loading-context'
...
-    image: placeholderImage,
+    id: 11,
+    title: 'Innovative Product',
+    price_sgd: '250',
+    description: 'A groundbreaking product that changes the game',
+    image_url: placeholderImageVerticallyLong,
+    number_sold: 5,
+    category: 'Technology',
+    link: 'https://www.example.com',
   },
```
Comment (nknguyenhc) [inline frontend/src/hooks/useProducts.ts]: You may want to remove this array altogether, since we are not using it anymore.

File: main.py [hunk 3]
```diff
@@ -3,11 +3,12 @@
 from dotenv import load_dotenv
 from fastapi import FastAPI, Request
 from fastapi.exceptions import RequestValidationError
-from fastapi.responses import FileResponse, JSONResponse
+from fastapi.responses import FileResponse, JSONResponse, RedirectResponse
 from fastapi.staticfiles import StaticFiles
 import os
 from starlette.config import Config
 from starlette.middleware.sessions import SessionMiddleware
+from starlette.middleware.cors import CORSMiddleware
```
Comment (nknguyenhc) [inline main.py]: Please remove unused imports

**Review by nknguyenhc — CHANGES_REQUESTED** [review]
A few issues here:

### General

1. Please remove all hardcoded origins `http://localhost:3000` and `http://localhost:8000` from functional code. Add them to env variables instead, so that it can be easily changed during deployment.

### Frontend

image

1. Please clear the warnings in the terminal.

image

2. The product card is quite long for medium viewport width (~950px), you may want to:

* Adjust the number of cards.
* Display only ~200 characters of the description, have a button so that the user only sees full description upon pressing it.
* Reduce the font size of product title for smaller screensizes.

image

3. The cookie can reveal info about the user, especially when user shares URL. You may want to remove the token from URL once the page has saved it to session storage.

4. You may want to keep the prompt in the input text so that the user can refer to what he/she prompted.

5. Please add a hyperlink to each card so that the user can be redirect to tiktok app, using the `link` prop.

File: frontend/src/context/auth-context.tsx (hunk 1 above)
Comment (Arif-Khalid) [reply frontend/src/context/auth-context.tsx]: Removed redundant typecast

File: frontend/src/hooks/useProducts.ts (hunk 2 above)
Comment (Arif-Khalid) [reply frontend/src/hooks/useProducts.ts]: Array removed

File: main.py (hunk 3 above)
Comment (Arif-Khalid) [reply main.py]: Removed unused imports

**Comment by Arif-Khalid** [comment]
I have replaced hardcoded url prefixes with the url within an env file and url no longer is used to transfer token after change made where cookie can be set in redirection response.
Points 2, 4, and 5 are currently being worked on by @freddychenyouren2

File: main.py [hunk 4]
```diff
@@ -48,48 +51,60 @@
...
+    return res
 
 
+@app.get("/token")
+async def token(request: Request):
+    auth_response = await oauth.auth0.authorize_access_token(
+        request, audience=os.getenv("AUTH0_AUDIENCE")
+    )
+    redRes = RedirectResponse(
+        url="http://127.0.0.1:3000/home/", status_code=302
```
Comment (nknguyenhc) [inline main.py]: URL here is still hardcoded

File: frontend/src/components/content-component/product-card.tsx [hunk 5]
```diff
@@ -1,42 +1,99 @@
...
+    if (!price) return "Contact Seller for Price";
+    const prices = price.split(' - ');
+    return prices.length === 2
+      ? `$${prices[0]} - $${prices[1]}`
+      : `$${prices[0]}`;
+  };
+
+  const handleRedirect = () => {
+    window.open(link)
+  }
```
Comment (nknguyenhc) [inline frontend/src/components/content-component/product-card.tsx]: You may want to:
1. Wrap constant definitions in `useMemo` hooks.
2. Wrap function definitions in `useCallback` hooks.

These reduce the number of definition refresh upon re-render and hence better performance. Constant/function declarations without hooks by default are refreshed on every re-render.

File: frontend/src/routes/private-route.tsx [hunk 6]
```diff
@@ -3,8 +3,11 @@ import { useAuth } from '../context/auth-context'
 
 export function PrivateRoute() {
   const user = useAuth()
-  if (!user || !user.token) {
+  if (!user) {
     return <Navigate to="/login" />
   }
+
+  user.authenticate()
+
```
Comment (nknguyenhc) [inline frontend/src/routes/private-route.tsx:10]: Please put functional code inside hooks, in this case probably a `useEffect`?

**Review by nknguyenhc — CHANGES_REQUESTED** [review]
image

I am unable to login. No matter how many times I press login, I will still be redirected back to login page.

File: main.py (hunk 4 above)
Comment (Arif-Khalid) [reply main.py]: Removed hard coded values.

File: frontend/src/components/content-component/product-card.tsx (hunk 5 above)
Comment (Arif-Khalid) [reply frontend/src/components/content-component/product-card.tsx]: @freddychenyouren2

File: frontend/src/routes/private-route.tsx (hunk 6 above)
Comment (Arif-Khalid) [reply frontend/src/routes/private-route.tsx:10]: Updated appropriately

**Pull request merged**